db/view/view.cc

```diff
@@ -1109,102 +1124,63 @@ future<> mutate_MV(
                 }
             }
-        } else if (!pending_endpoints.empty()) {
-            // If there is no paired endpoint, it means there's a range movement going on (decommission or move),
-            // such that this base replica is gaining new token ranges. The current node is thus a pending_endpoint
-            // from the POV of the coordinator that sent the request. Since we only look at natural endpoints to
-            // determine base-to-view pairings, the current node won't appear in the list of base replicas. Sending
-            // view updates to the view replica this base will eventually be paired with only makes a difference when
-            // the base update didn't make it to the node which is currently being decommissioned or moved-from. Also,
-            // if HH is enabled at the coordinator, the update will either make it there before the range movement
-            // finishes, or later to this node when it becomes a natural endpoint for the token. We still ensure we
-            // send to any pending view endpoints though.
-            auto updates_pushed_remote = pending_endpoints.size();
+        }
+
+        if (target_endpoint && *target_endpoint == my_address) {
+            ++stats.view_updates_pushed_local;
+            ++cf_stats.total_view_updates_pushed_local;
+            ++stats.writes;
+            auto mut_ptr = remote_endpoints.empty() ? std::make_unique<frozen_mutation>(std::move(mut.fm)) : std::make_unique<frozen_mutation>(mut.fm);
+            tracing::trace(tr_state, "Locally applying view update for {}.{}; base token = {}; view token = {}",
+                    mut.s->ks_name(), mut.s->cf_name(), base_token, view_token);
+            future<> local_view_update = service::get_local_storage_proxy().mutate_locally(mut.s, *mut_ptr, std::move(tr_state), db::commitlog::force_sync::no).then_wrapped(
+                    [&stats,
+                     maybe_account_failure = std::move(maybe_account_failure),
+                     mut_ptr = std::move(mut_ptr)] (future<>&& f) {
+                --stats.writes;
+                return maybe_account_failure(std::move(f), utils::fb_utilities::get_broadcast_address(), true, 0);
+            });
+            fs->push_back(std::move(local_view_update));
+            // We just applied a local update to the target endpoint, so let's choose another
+            // primary target for the remote update - if there are any remote endpoints to choose from.
+            if (!remote_endpoints.empty()) {
+                target_endpoint = std::move(remote_endpoints.back());
+                remote_endpoints.pop_back();
+            } else {
+                target_endpoint.reset();
+            }
+        }
+
+        // If paired_endpoint is engaged by this point, then either the update
+        // is not local, or the local update was already applied but we still
+        // have pending endpoints to send to.
```

Comment:
Is it possible that the update is not local *and* there is a pending endpoint as well?

Reply:
If I understand the notion of pending endpoints correctly, it's orthogonal to the locality of a view update. So if we send an update to a replica which has a paired endpoint, but this paired endpoint is in the middle of range movement, then it can have a pending endpoint as well.

Reply:
So in that case, IIUC, the code will skip sending the update to the pending endpoint. The first if above will be skipped as the target is remote, and here we will send the update to just one endpoint.

Reply:
`send_to_endpoint` is a function that takes one target endpoint and a vector of pending endpoints - so it will send the update to one "target" endpoint, since the interface obliges us to pick a single special endpoint to be a "target", but it will also send to all pending endpoints (`remote_endpoints`).

Reply:
I see.

Reply:
Oh, do you mean the case when there's **no** paired endpoint, but there are pending endpoints? I'm not sure if it's possible (and that also means I have no idea how to reach such scenario in tests), but I think the previous code handled this case, so you're right, I'll fix this pull request to handle this case.

Reply:
It's not what I meant, but I'm glad my ignorant comment made you think of a problem. :)